Replace javascript highlighting in guides with Rouge [ci skip]

p8 opened the pull request:
### Summary

Replace javascript highlighting using Prism with the Rouge markdown highlighter.
Uses the github theme with some adjustments.
Follow up to #37972 
cc @robin850 @kaspth 

### Before
<img width="537" alt="image" src="https://user-images.githubusercontent.com/28561/86390085-b28ead80-bc97-11ea-8e4a-209de46d9cb7.png">
<img width="560" alt="image" src="https://user-images.githubusercontent.com/28561/86390158-dc47d480-bc97-11ea-9a3d-10e0bf739576.png">
<img width="549" alt="image" src="https://user-images.githubusercontent.com/28561/86390287-15804480-bc98-11ea-81e7-85949f524e62.png">
<img width="552" alt="image" src="https://user-images.githubusercontent.com/28561/86390415-4ceef100-bc98-11ea-8621-60cb318a88b3.png">


### After
<img width="599" alt="image" src="https://user-images.githubusercontent.com/28561/86390049-9be85680-bc97-11ea-9beb-b756d416fb60.png">
<img width="619" alt="image" src="https://user-images.githubusercontent.com/28561/86390785-0057e580-bc99-11ea-9114-ee4b0a047406.png">

<img width="607" alt="image" src="https://user-images.githubusercontent.com/28561/86391029-66446d00-bc99-11ea-81d8-84cee3684c09.png">

<img width="601" alt="image" src="https://user-images.githubusercontent.com/28561/86390384-3e083e80-bc98-11ea-9fde-10ac249f6400.png">



### Other Information
I've tried to also keep "shell command selection" working (#39594).
It seems to work except for the whitespace in front of the selection:
<img width="604" alt="image" src="https://user-images.githubusercontent.com/28561/86389628-e1f0ea80-bc96-11ea-9b4a-5078973f64b6.png">

jonathanhefner commented on guides/rails_guides/markdown/renderer.rb:
```diff
@@ -1,14 +1,22 @@
 
       def block_code(code, language)
+        formatter = Rouge::Formatters::HTML.new
+        lexer = ::Rouge::Lexer.find_fancy(lexer_language(language), code)
```
I think the `code` arg here is only used when [specifying `"guess"` as the language](https://github.com/rouge-ruby/rouge/blob/v3.20.0/lib/rouge/lexer.rb#L47-L53).

p8 replied:
I've removed the `code` arg

jonathanhefner commented on guides/source/routing.md:
````diff
@@ -24,7 +24,7 @@ The Rails router recognizes URLs and dispatches them to a controller's action, o
 When your Rails application receives an incoming request for:
 
-```
+```plain
````
What will be the difference between `plain` and unmarked code fences?

p8 replied:
They will be similar. At first I tried using the `http` lexer but these require the protocol as well for the requests.
```http
"GET /photos/7 HTTP/1.1"
```
I'll remove these changes..

jonathanhefner commented on guides/assets/stylesheets/main.css:
```diff
@@ -729,12 +729,9 @@ div.important p, div.caution p, div.warning p, div.note p, div.info p {
 
 /* When triple-clicking shell code, select only the command, not the prompt */
-code.language-shell-session span.token.command {
+code.highlight.console span.gp {
   display: inline-flex;
 }
```
Try this:

```suggestion
code.highlight.console span.w {
  display: table-cell;
}
```

Or, if `.w` shows up in more places and we need more specificity:

```suggestion
code.highlight.console span.gp + span.w {
  display: table-cell;
}
```

I tested in Firefox and Chromium, and triple-click selection from #39594 works for recognized commands (`BUILTINS`).  It doesn't work as well for non-recognized commands because Rouge doesn't add a `span` after the `.w` in that case, but that's probably okay.

p8 replied:
Thanks! I've chosen the simpler first solution. I've grep-ed for `.w` and it only seems to be used for the console commands.

jonathanhefner reviewed (commented):
Replacing JavaScript with Ruby always gets a vote from me! :grin:

p8 commented on guides/rails_guides/markdown/renderer.rb:
```diff
@@ -1,14 +1,22 @@
+require "rouge"
+
+# Add more common shell commands
+Rouge::Lexers::Shell::BUILTINS += "|bin/rails|brew|bundle|gem|git|node|rails|rake|ruby|sqlite3|yarn"
```
Fixed.

jonathanhefner commented on guides/source/routing.md:
```diff
@@ -592,7 +592,7 @@ TIP: By default, dynamic segments don't accept dots - this is because the dot is
 ### Static Segments
 
-You can specify static segments when creating a route by not prepending a colon to a segment:
+You can specify static segments when creating a route by not prepending a colon to a fragment:
```
It looks like the changes in this file got picked up from #39775?

p8 replied:
Ah thanks 👍 . I've removed these changes.

p8 committed:
Use Rouge for guides code highlighting [ci skip]

Replace javascript highlighting using Prism with the Rouge markdown
highlighter. Uses the github theme.

eugeneius commented on guides/rails_guides/markdown/renderer.rb:
```diff
@@ -58,18 +66,20 @@ def convert_footnotes(text)
             "erb"
           when "html"
-            "xml" # HTML is understood, but there are .xml rules in the CSS
+            "html"
```
This is already covered by the first branch.

p8 replied:
I've removed this branch

eugeneius commented on guides/rails_guides/markdown/renderer.rb:
```diff
@@ -58,18 +66,20 @@ def convert_footnotes(text)
           when "bash"
-            "shell-session"
+            "console"
+          when "plain"
+            "plaintext"
```
Could this be ✂️, if it's the same as the default case?

p8 replied:
Fixed.

eugeneius commented:
There's a lot more padding around code blocks after this change:

<img width="200" alt="Screenshot 2020-07-05 at 21 24 43" src="https://user-images.githubusercontent.com/432189/86541582-f71f8080-bf05-11ea-85f5-f0d400b768a4.png"> <img width="200" alt="Screenshot 2020-07-05 at 21 23 43" src="https://user-images.githubusercontent.com/432189/86541570-d6efc180-bf05-11ea-929b-0e380536cdbf.png">

I think we can fix it by updating the main stylesheet here:

https://github.com/rails/rails/blob/2f87589630aa30099fb8f910bd3963ff95be7e23/guides/assets/stylesheets/main.css#L32

to match the margin that was previously applied by Prism:

https://github.com/rails/rails/blob/2f87589630aa30099fb8f910bd3963ff95be7e23/guides/assets/stylesheets/prism/default.css#L55

p8 commented:
@eugeneius, good catch!.

On edge there is a margin of 0.5em but it cuts off the image:
<img width="645" alt="image" src="https://user-images.githubusercontent.com/28561/86560970-c9ffbc00-bf5f-11ea-937b-387d4e93bdf4.png">

This was broken by the Prism change because on stable there seems to be a margin of 1em:
<img width="645" alt="image" src="https://user-images.githubusercontent.com/28561/86560900-a2105880-bf5f-11ea-85f1-1a74c26628f1.png">

I've changed the margin to 1em:
<img width="650" alt="image" src="https://user-images.githubusercontent.com/28561/86561286-9709f800-bf60-11ea-95ba-9d22a3199f43.png">

eugeneius commented on guides/rails_guides/markdown/renderer.rb:
```diff
@@ -58,18 +66,16 @@ def convert_footnotes(text)
+        def lexer_language(code_type)
           case code_type
-          when "ruby", "sql", "plain", "js", "yaml"
+          when "js", "http", "html", "ruby", "sql", "yaml", "yml"
```
One last nitpick, "http" is unused:

```suggestion
          when "js", "html", "ruby", "sql", "yaml", "yml"
```

p8 replied:
I've removed `http`.
I've also removed `yml` because it was used only once, and I've replaced the code block to use `yaml` instead.

p8 replied:
Sorry, something went wrong with pushing changes. I've pushed the above mentioned changes.

eugeneius commented:
Thanks @p8, this looks great!

jonathanhefner commented:
Thank you @p8!  Thank you @eugeneius!

:heart:

robin850 commented:
Awesome, really happy with this, thanks folks ! ❤️
